data/nat/choose.lean

```diff
@@ -15,83 +15,100 @@ def choose : ℕ → ℕ → ℕ
-  cases n with n,unfold choose,unfold choose,
-end
+lemma choose_zero_right (n : ℕ) : choose n 0 = 1 := by cases n;refl
 
 @[simp]
 lemma choose_zero_succ (k : ℕ) : choose 0 (succ k) = 0 := rfl
 
 lemma choose_succ_succ (n k : ℕ) : choose (succ n) (succ k) = choose n k + choose n (succ k) := rfl
 
-lemma choose_eq_zero_of_lt {n k : ℕ} : n < k → choose n k = 0 := begin
-  revert k,
-  induction n with n' hi,
-  assume k nk,cases k,exact absurd nk dec_trivial,
-  unfold choose,
-  assume k nk,
-  cases k with k',
-  exact absurd nk dec_trivial,
-  unfold choose,
-  rw [@hi k' (lt_of_succ_lt_succ nk),@hi (k' + 1) (lt_of_succ_lt nk)],
+lemma choose_eq_zero_of_lt {n : ℕ} : ∀ {k}, n < k → choose n k = 0 :=
+begin
+  induction n with n hi,
+  { assume k hk, cases k,
+    exact absurd hk dec_trivial,
+    exact choose_zero_succ _},
+  { assume k hk,
+    cases k with k',
+      {exact absurd hk dec_trivial},
+    unfold choose,
+    rw [@hi _ (lt_of_succ_lt_succ hk),@hi _ (lt_of_succ_lt hk)]}
 end
 
 @[simp]
-lemma choose_self (n : ℕ) : choose n n = 1 := begin
-  induction n with n' hi,simp!,simpa! [choose_eq_zero_of_lt (lt_succ_self _)],
+lemma choose_self (n : ℕ) : choose n n = 1 :=
+begin
+  induction n with n' hi, 
+  { simp [choose]},
+  { simpa! [choose_eq_zero_of_lt (lt_succ_self _)]}
 end
 
 @[simp]
 lemma choose_succ_self (n : ℕ) : choose n (succ n) = 0 := choose_eq_zero_of_lt (lt_succ_self _)
 
 @[simp]
-lemma choose_one_right (n : ℕ) : choose n 1 = n := begin
-  induction n with n' hi,unfold choose,unfold choose,simp[hi],
+lemma choose_one_right (n : ℕ) : choose n 1 = n :=
+begin
+  induction n with n' hi,
+  { refl},
+  { unfold choose,simp[hi]},
 end
 
-lemma choose_pos {n k : ℕ} : k ≤ n → 0 < choose n k := begin
-  revert k,
-  induction n with n' hi,assume k kn,
-  rw eq_zero_of_le_zero kn,simp,exact dec_trivial,
-  assume k kn,
-  cases k with k',simp,exact dec_trivial,
-  rw choose_succ_succ,
-  exact add_pos_of_pos_of_nonneg (hi (le_of_succ_le_succ kn)) (zero_le _),
+lemma choose_pos {n : ℕ} : ∀ {k}, k ≤ n → 0 < choose n k :=
+begin
+  induction n with n' hi,
+  { assume k hk,
+    rw [eq_zero_of_le_zero hk,choose_zero_right],
+    exact dec_trivial},
+  assume k hk,
+  cases k with k',
+  { simp, exact dec_trivial},
+  { rw choose_succ_succ,
+    exact add_pos_of_pos_of_nonneg (hi (le_of_succ_le_succ hk)) (zero_le _)},
 end
 
-lemma succ_mul_choose_eq (n k : ℕ) : succ n * choose n k = choose (succ n) (succ k) * succ k := begin
-  revert k,induction n with n' hi,assume k,
-  simp,cases k with k',simp,simp,
-  rw [choose_eq_zero_of_lt (dec_trivial : 1 < succ(succ k')),zero_mul],
-  assume k,cases k with k',
-  simp,
-  rw [choose_succ_succ (succ n') (succ k'),add_mul,←hi,mul_succ,←hi,add_right_comm],
-  rw [←mul_add,←choose_succ_succ,←succ_mul],
+lemma succ_mul_choose_eq (n : ℕ) : ∀ k, succ n * choose n k = choose (succ n) (succ k) * succ k := 
+begin
+  induction n with n' hi,
+  { assume k, simp,
+    cases k with k', 
+      simp,
```

Comment:
Put `simp` into `{ simp }`. This is especially important for `simp`. Otherwise it might happen that the simp set changes and the next tactics still work. Then it is hard to figure out which `simp` call failed.